src/telem/lib/input/mekanism/DynamicTankInputAdapter.lua

```diff
@@ -0,0 +1,68 @@
+local o = require 'telem.lib.ObjectModel'
+local t = require 'telem.lib.util'
+
+local InputAdapter      = require 'telem.lib.InputAdapter'
+local Metric            = require 'telem.lib.Metric'
+local MetricCollection  = require 'telem.lib.MetricCollection'
+
+local DynamicTankInputAdapter = o.class(InputAdapter)
+DynamicTankInputAdapter.type = 'DynamicTankInputAdapter'
+
+function DynamicTankInputAdapter:constructor (peripheralName, categories)
+    self:super('constructor')
+
+    -- TODO this will be a configurable feature later
+    self.prefix = 'mekchemtank:'
+
+    -- TODO make these constants
+    local allCategories = {
+        'basic',
+        'chemical'
+    }
+
+    if not categories then
+        self.categories = { 'basic' }
+    elseif categories == '*' then
+        self.categories = allCategories
+    else
+        self.categories = categories
+    end
+
+    -- boot components
+    self:setBoot(function ()
+        self.components = {}
+
+        self:addComponentByPeripheralID(peripheralName)
+    end)()
+end
+
+function DynamicTankInputAdapter:read ()
+    self:boot()
+
+    local source, tank = next(self.components)
+
+    local metrics = MetricCollection()
+
+    local loaded = {}
+
+    for _,v in ipairs(self.categories) do
+        -- skip, already loaded
+        if loaded[v] then
+            -- do nothing
+
+        -- Literally all we have lmao
+        elseif v == 'basic' then
```

Comment:
Recommend putting everything in basic category to match Chemical Tank adapter layout.

I think `tank_capacity` and `chemical_tank_capacity` would be the most correct here. Dynamic Tanks can store more mB of chemical vs fluid for the same structure dimensions, so there are technically two maximum capacities depending on the contents. 🤷